File: client/src/main/scala/stasis/client/model/SourceEntity.scala
package stasis.client.model

final case class SourceEntity(
  ref: EntityRef,
  existingMetadata: Option[EntityMetadata],
  currentMetadata: EntityMetadata
) {
  existingMetadata.foreach(existing =>
    require(
      existing.getClass == currentMetadata.getClass,
      s"Mismatched current metadata for [${currentMetadata.path}] and existing metadata for [${existing.path}]"
    )
  )

  lazy val hasChanged: Boolean =
    existingMetadata match {
      case Some(existing) => existing.hasChanged(comparedTo = currentMetadata)
      case None           => true
    }

  lazy val hasContentChanged: Boolean =
    (existingMetadata, currentMetadata) match {
      case (Some(existing: EntityMetadata.WithContent), current: EntityMetadata.WithContent) =>
        existing.size != current.size || existing.checksum != current.checksum

      case (None, _: EntityMetadata.WithContent) =>
        true

      case _ =>
        false
    }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/model/SourceEntity.kt
package stasis.client_android.lib.model

data class SourceEntity(
    val ref: EntityRef,
    val existingMetadata: EntityMetadata?,
    val currentMetadata: EntityMetadata
) {
    init {
        existingMetadata?.let { existing ->
            require(existing.javaClass == currentMetadata.javaClass) {
                "Mismatched current metadata for [${currentMetadata.path}] and existing metadata for [${existing.path}]"
            }
        }
    }

    val hasChanged: Boolean by lazy {
        when (existingMetadata) {
            null -> true
            else -> existingMetadata.hasChanged(comparedTo = currentMetadata)
        }
    }

    val hasContentChanged: Boolean by lazy {
        val current = currentMetadata
        val existing = existingMetadata
        when {
            existing is EntityMetadata.WithContent && current is EntityMetadata.WithContent ->
                existing.size != current.size || existing.checksum != current.checksum

            existing == null && current is EntityMetadata.WithContent -> true

            else -> false
        }
    }
}
